tensorflow/core/user_ops/gather_columns.cc

```diff
@@ -0,0 +1,183 @@
+#include "tensorflow/core/framework/op.h"
+//#include "tensorflow/core/framework/shape_inference.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/platform/prefetch.h"
+
+using namespace tensorflow;
+using namespace std;
+
+
+REGISTER_OP("GatherColumns")
+    .Input("params: T")
+    .Input("indices: Index")
+    .Output("columns: T")
+    .Attr("T: type")
+    .Attr("Index: {int32,int64}");
+
+template <typename T>
+class GatherColumnsOp : public OpKernel {
+ public:
+  explicit GatherColumnsOp(OpKernelConstruction* ctx) : OpKernel(ctx) {}
+
+  void Compute(OpKernelContext* ctx) override {
+
+    //--Grab the input tensor - params--//
+    const Tensor& params = ctx->input(0);
+
+    //--Grab the input tensor - indices--//
+    const Tensor& indices = ctx->input(1);
+    auto ind_flat = indices.flat<int32>();
+
+    OP_REQUIRES(ctx, TensorShapeUtils::IsVectorOrHigher(params.shape()),
+                errors::InvalidArgument("params must be at least a vector"));
+    OP_REQUIRES(ctx, TensorShapeUtils::IsVector(indices.shape()),
+                errors::InvalidArgument("indices must be a vector, but it a: ", indices.dims(), "D Tensor."));
+
+    const TensorShape& params_shape(params.shape());
+
+    OP_REQUIRES(
+        ctx, params_shape.dims() <= 2,
+        errors::InvalidArgument("params must be 1D or 2D but it is: ", params_shape.dims(), "D"));
+
+    TensorShape output_shape(params_shape);
+
+    int64 params_rows;
+    int64 params_cols;
+    int64 indices_size= indices.dim_size(0);
+
+    OP_REQUIRES(ctx, indices_size > 0,
+                errors::InvalidArgument("indices cannot be a empty."));
+
+    if(params_shape.dims() == 1)
+    {
+      params_rows = 1;
+      params_cols = params.dim_size(0);
+
+      //--Set output tensor dims--//
+      output_shape.set_dim(0, indices_size);
+    }
+    else if(params_shape.dims() == 2)
+    {
+      params_rows = params.dim_size(0);
+      params_cols = params.dim_size(1);
+
+      //--Set output tensor dims--//
+      output_shape.set_dim(0, params_rows);
+      output_shape.set_dim(1, indices_size);
+    }
+
+    //--Check indices[i] ∈ (0, params_cols], ∀i--//
+    for(int64 i=0; i < indices_size; i++)
+    {
+      //--TODO: Should look for a more optimal way to do this, or maybe there is a TF macro for this--//
+      OP_REQUIRES(
+        ctx, ind_flat(i) >= 0 && ind_flat(i) < params_cols,
+        errors::InvalidArgument("indices(", i, "): ", ind_flat(i), " is not in range (0, ", params_cols, "]."));
+    }
+
+    //--Create an output tensor--//
+    Tensor* output_tensor = NULL;
+    OP_REQUIRES_OK(ctx, ctx->allocate_output(0, output_shape, &output_tensor));
+
+    //--Single column tensor, indices must include it, so just copy param tensor to output tensor--//
+    if(params_cols == 1)
```

Comment:
Not sure if it's possible, but I need to look into it. 

Let me do this post GPU implementation, through an independent push.